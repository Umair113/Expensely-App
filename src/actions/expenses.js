import uuid from 'uuid';
import database from '../firebase/firebase'
/*
    component calls action generator
    action generator will return a function
    component dispatches function
    function runs - has the ability ot dispatch other actions and do whatever it wants
*/

// ADD EXPENSE
export const addExpense = ( expense) => {
    return {
        type: 'ADD_EXPENSE',
        expense
    };
};


// returns the function for dispatch
export const startAddExpense = (expenseData={}) => {
    return (dispatch) => {
        const {
            description='', 
            note='', 
            amount=0, 
            createdAt=0
        } = expenseData;

        const expense = {description, note, amount, createdAt};

        // return this so we can attach a promise chain for the test case
        return database.ref('expenses').push(expense)
            .then( (ref)=> {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    };
};


// REMOVE EXPENSE
export const removeExpense = ( {id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id}={}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({
                    id
                }));
            });
    };
};

// EDIT EXPENSE
export const editExpense = ( id, updates ) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch => {
        const expensePath = `expenses/${id}`;
        return database.ref(expensePath).update(
            updates
        ).then(()=>{
            dispatch(editExpense(id, updates));
        });
    });
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach(childSnapshot => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            
            dispatch(setExpenses(expenses));
        });
    };
};
