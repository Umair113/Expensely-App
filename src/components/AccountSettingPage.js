import React from 'react';

const AccountSettingPage = () => (
    <div className="content-container">
        <h1>Account Settings</h1>
        In progress!

        <ul>
            <li>Manage Features</li>
            <ul>
                <li>Expense setting (change curreny?)</li>
                <li>Goal settings?</li>
            </ul>
            <li>Bank Accounts</li>
            <li>Security</li>
            <li>Notifications</li>
            <li>Profile</li>
        </ul>


    </div>
);

export default AccountSettingPage;