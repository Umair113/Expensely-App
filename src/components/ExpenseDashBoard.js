import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashBoardPage = () => (
    <div>
        <ExpenseListFilters/>
        <ExpenseSummary/>
        <ExpenseList/>
    </div>
);

export default ExpenseDashBoardPage;