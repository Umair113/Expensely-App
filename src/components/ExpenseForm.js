import React from 'react';
import moment from 'moment';
import {connect} from 'react-redux';
import {SingleDatePicker} from 'react-dates';

export class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            errorMessage: ''
        };
    }
    
    onDescriptionChange = (e) =>{
        const description = e.target.value;
        this.setState(()=> ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState( () => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;

        if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt){
            this.setState( ()=> ({createdAt}));
        }
    };
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));
    };

    componentDidCatch(error, info) {
        console.log('Error occured:',error, info);
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount){
            // error handling: provide description and amount
            this.setState(()=> ({errorMessage : 'Amount and description fields are required.'}));
        } else {
            this.setState(()=> ({errorMessage : ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    onRemove = (e) =>{
        this.props.onRemove({
            id: this.props.id
        })
    };

    render () {
        return (
            <div className="content-container">
                
                {this.state.errorMessage && <p>{this.state.errorMessage}</p>}
                <form className="form" onSubmit={this.onSubmit}>
                    <span>Description</span>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        className="text-input"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <span>Amount</span>
                    <input
                        type="text"
                        placeholder="Amount"
                        className="text-input"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <span>Date</span>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <span>Note</span>
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        className="textarea"
                        value={this.state.note}
                        onChange = {this.onNoteChange}
                    >
                    </textarea>
                    <div className="button-bar-Container">
                        <button type="submit" className="button">Save Expense</button>
                        {!!this.props.onRemove && <button type="button" className="button" onClick={this.onRemove}>Remove</button>}
                        <button type="button" className="button" onClick={this.props.onCancel}>Cancel</button>
                    </div>
                </form>
            </div>
        )
    }
}

export default ExpenseForm;


// const mapStateToProps = (state, props) => {
//     return{
//         expense: state.expenses.find((expense) => expense.id === props.match.params.id)
//     };
// };

// const mapDispatchToProps = (dispatch, props) => {
//     return {
//         startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
//     };
// };
// export default connect(undefined, mapDispatchToProps)(ExpenseForm);