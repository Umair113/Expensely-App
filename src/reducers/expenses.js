
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type){
        case 'ADD_EXPENSE':
            // Array Spread operator. Makes it easy to concat values into an array and get a new array 
            // * doesn't affect the old array which is what we WANT
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter( ({id}) => action.id !== id);
        case 'EDIT_EXPENSE':
            return state.map( (expense)=> {
                if (expense.id === action.id){
                    // object Spread operator. Makes it easy to combine objects/object elements and get a new object returned 
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                  return expense;
                };
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};