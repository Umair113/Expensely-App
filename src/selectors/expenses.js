import moment from 'moment';

// Get visible expense - apply filters to expenses
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter( (expense) => {
        
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

        const textMatch = expense.description.toUpperCase().includes(text.toUpperCase());

        return startDateMatch && endDateMatch && textMatch;

    }).sort( (a,b) => {
        if (sortBy === 'date') {
            // if return value is less than 0, a is a lower index than b. a comes first
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else if (sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        } else {
            console.log('unknown sort');
        }
    });
};

export default getVisibleExpenses;