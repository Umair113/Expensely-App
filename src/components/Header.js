import React from 'react';
import { NavLink } from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';

export const Header = ({startLogout}) => (
    <header className="header">
        <div className ="header-container">
            <div className="content-container">
                <h1>Expense-Ly</h1>
            </div>
        </div>

        <div className='nav-toolbox-separator'>
            <div className="content-container">
                <div className="nav-container">
                    <div className="nav-links">
                        <NavLink to="/dashboard" className = 'nav-item' activeClassName="is-active" >Dashboard</NavLink>
                        <NavLink to="/help" className = 'nav-item' activeClassName="is-active" >Help</NavLink>
                    </div>
                    <div className = "nav-logout">
                        <button className='nav-item' onClick={startLogout}>Logout</button>
                    </div>
                </div>
            </div>
        </div>

    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
