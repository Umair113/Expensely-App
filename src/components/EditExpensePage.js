import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// refactored editexpensepage to be a class based component
// setup mapDispatchToProps

// should render editExpensepage
// snapshot

// should handle editExpense
// spies

// should handle removeExpense
// spies

export class EditExpensePage extends React.Component {
    startEditExpense = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };

    // class method to render
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.startEditExpense} 
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>
                        Remove
                    </button>
                </div>
            </div>
        );
    };
}

const mapStateToProps = (state, props) => {
    return{
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
