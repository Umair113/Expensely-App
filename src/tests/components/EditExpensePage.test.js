import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
    startEditExpense = jest.fn(); 
    startRemoveExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow (
        <EditExpensePage
            startEditExpense={startEditExpense}
            startRemoveExpense={startRemoveExpense}
            history={history}
            expense={expenses[2]}
        />
    )
});


// should render editExpensepage
test('should render editExpensePage', () => {
    expect(wrapper).toMatchSnapshot();
});

// should handle editExpense
test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

// // should handle removeExpense
test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({
        id: expenses[2].id
    });
});