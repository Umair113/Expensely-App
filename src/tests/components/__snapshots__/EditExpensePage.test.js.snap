// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 5034,
          "createdAt": 345600000,
          "description": "credit card",
          "id": "3",
          "note": "citi bank",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
